test/Microsoft.AspNet.Mvc.FunctionalTests/XmlDataContractSerializerInputFormatterTest.cs

```diff
@@ -0,0 +1,133 @@
+            Assert.Equal(expectedErrorMessages.Count(), modelBindingInfo.ModelStateErrorMessages.Count);
+            foreach (var expectedErrorMessage in expectedErrorMessages)
+            {
+                Assert.Contains(
+                modelBindingInfo.ModelStateErrorMessages,
+                (actualErrorMessage) => actualErrorMessage.StartsWith(expectedErrorMessage));
+            }
+        }
+
+        // Verifies that the model state has errors related to body model validation(for reference types) and also for
+        // Required attribute validation (for value types).
+        [Fact]
+        public async Task DataMissingForReferneceTypeProperties_AndModelIsBound_AndHasMixedValidationErrors()
+        {
+            // Arrange
+            var server = TestServer.Create(_services, _app);
+            var client = server.CreateClient();
+            client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/xml-dcs"));
+            var input = "<Store xmlns=\"http://schemas.datacontract.org/2004/07/XmlFormattersWebSite\"" +
+                        " xmlns:i=\"http://www.w3.org/2001/XMLSchema-instance\"><Address><State i:nil=\"true\"/>" +
+                        "<Zipcode>98052</Zipcode></Address><Customers><Customer><Address i:nil=\"true\"/><Id>1000" +
+                        "</Id></Customer></Customers><Id>10</Id></Store>";
+            var content = new StringContent(input, Encoding.UTF8, "application/xml-dcs");
+            var propertiesCollection = new List<KeyValuePair<string, string>>();
+            propertiesCollection.Add(new KeyValuePair<string, string>(nameof(Store.Id), typeof(Store).FullName));
+            propertiesCollection.Add(new KeyValuePair<string, string>(nameof(Address.Zipcode), typeof(Address).FullName));
+            propertiesCollection.Add(new KeyValuePair<string, string>(nameof(Customer.Id), typeof(Customer).FullName));
+            var expectedErrorMessages = propertiesCollection.Select(kvp =>
+            {
+                return string.Format(
+                    "{1}:Value type property '{0}' on type '{1}' has RequiredAttribute but " +
+                    "no DataMember(IsRequired = true) attribute.",
+                    kvp.Key, kvp.Value);
+            }).ToList();
+            expectedErrorMessages.Add("store.Address.State:The State field is required.");
+            expectedErrorMessages.Add("store.Customers[0].Address:The Address field is required.");
+
+            // Act
+            var response = await client.PostAsync("http://localhost/Validation/CreateStore", content);
+
+            //Assert
+            var dcsSerializer = new DataContractSerializer(typeof(ModelBindingInfo));
+            var responseStream = await response.Content.ReadAsStreamAsync();
+            var modelBindingInfo = dcsSerializer.ReadObject(responseStream) as ModelBindingInfo;
+            Assert.NotNull(modelBindingInfo);
```

Comment:
nit: with this huge set of validation, I find it hard to fish what is actually being validated. We could just had two reference type and two value types and the validation would have been simpler.